lib/line-numbers.js

```diff
@@ -0,0 +1,118 @@
+	constructor(offender) {
+		const numberPluralized = pluralizeIf('number', offender && !NUMBER_REGEX.test(offender) && offender.length !== 1);
+		super(`Invalid line ${numberPluralized}: \`${offender}\`. Expected comma-separated list of \`[X|Y-Z]\`.`);
+		this.name = 'FormatError';
+	}
+}
+
+const distinctArray = array => [...new Set(array)];
+const sortNumbersAscending = array => {
+	const sorted = [...array];
+	sorted.sort((a, b) => a - b);
+	return sorted;
+};
+
+const parseNumber = string => parseInt(string, 10);
+const removeAllWhitespace = string => string.replace(/\s/g, '');
+const range = (start, end) => new Array(end - start + 1).fill(start).map((element, index) => element + index);
+
+const parseLineNumbers = suffix => sortNumbersAscending(distinctArray(flatten(suffix.split(',').map(part => {
+	if (NUMBER_REGEX.test(part)) {
+		const number = parseNumber(part);
+		if (number <= 0) {
+			throw new Error(`Invalid line number: \`${part}\`. Line numbers must be positive.`);
+		}
+
+		return number;
+	}
+
+	const match = RANGE_REGEX.exec(part);
+	if (match) {
+		const {startGroup, endGroup} = match.groups;
+		const start = parseNumber(startGroup);
+		const end = parseNumber(endGroup);
+		if (start < 0 || end < 0) {
+			throw new Error(`Invalid line number range: \`${part}\`. Line numbers must be positive.`);
+		}
+
+		if (start > end) {
+			throw new Error(`Invalid line number range: \`${part}\`. \`start\` must be less than \`end\`.`);
+		}
+
+		return range(start, end);
+	}
+
+	throw new FormatError(part);
+}))));
+
+function splitPatternAndLineNumbers(pattern) {
+	const parts = pattern.split(DELIMITER);
+	if (parts.length === 1) {
+		return {pattern, lineNumbers: null};
+	}
+
+	const suffix = removeAllWhitespace(parts.pop());
+	if (!LINE_NUMBERS_REGEX.test(suffix)) {
+		return {pattern, lineNumbers: null};
+	}
+
+	return {pattern: parts.join(DELIMITER), lineNumbers: parseLineNumbers(suffix)};
+}
+
+exports.splitPatternAndLineNumbers = splitPatternAndLineNumbers;
+
+function getApplicableLineNumbers(normalizedFilePath, filter) {
+	return sortNumbersAscending(distinctArray(flatten(
+		filter
+			.filter(({pattern}) => micromatch.isMatch(normalizedFilePath, pattern))
+			.map(({lineNumbers}) => lineNumbers)
+	)));
+}
+
+exports.getApplicableLineNumbers = getApplicableLineNumbers;
+
+const resolveEndLineNumberForTestInFile = ({title, startLineNumber}, filePath) => {
+	try {
+		const node = parseTestSourceInFile({title, startLineNumber}, filePath);
+		return node.loc.end.line;
+	} catch (error) {
+		throw new Error(`Failed to resolve end line number for test \`${title}\` starting at line number ${startLineNumber} in ${filePath}: ${error.message}`);
+	}
+};
+
+function getLineNumberRangeForTestInFile(title, filePath) {
+	const callSite = callsites().find(callSite => callSite.getFileName() === filePath);
+	if (!callSite) {
+		throw new Error(`Failed to resolve line number range for test \`${title}\` in ${filePath}.`);
```

Comment:
When would this happen?

Once you start declaring tests, you must declare them all at once, but you could wait before declaring the first test. So I suppose you could have the actual test declarations in a separate file, and then use a `setTimeout()`, and then we wouldn't be able to find any call site.

And if that's not the issue, but the declarations are still happening in a different file, then line number selection is just not going to work.

I wonder if instead of throwing an error, we should print a warning and just run all tests. Here and when resolving the end line. What do you think?

Reply:
Hehe, that error has re-surfaced quite a few times now...

So, it's not an end user error. It's a developer error. If the function `getLineNumberRangeForTestInFile` was to be used/called without a test declaration being the initiator, in which case there would be no relevant call site to be found.

Since I've learned you're not really fond of DX errors, I think we can just drop it altogether.

Reply:
Yes, but you're assuming tests are declared in the test file. That would be typical, but it's not necessary.

Rather than failing, maybe we should log warnings and run whatever tests we did manage to select by line number. And if it turns out that we couldn't match any, then at least there's warnings.

Reply:
Wasn’t aware of that use case, but sure. Makes sense.

Reply:
Uhm, how would you print a warning from here?

Reply:
I think either a warning if no tests were selected, or perhaps a warning if not all lines / ranges lead to a test being selected. But that wouldn't be here, that'd be something in the runner maybe. I can give this a go if you'd like.

(It would also help us deal with Babel & TypeScript… we'd just warn that we couldn't find tests for now.)

Reply:
Yeah, sure. I was more thinking how do you normally (technically) communicate warnings from the non-UI code? Isn't that done in the reporters like with the other case we had. When I come to think about it, isn't it pretty much the same warning/error?

True that about Babel & TS.

Reply:
I… don't immediately have an answer, sorry. I can look into this.